hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/CatalogReplicaReplicationEndpoint.java

```diff
@@ -0,0 +1,493 @@
+      CatalogReplicaShipper[] replicaShippers = primaryRegionsUnderReplication.get(entry.getKey());
+      if (replicaShippers != null) {
+        for (List<Entry> l : entry.getValue()) {
+          for (CatalogReplicaShipper shipper : replicaShippers) {
+            shipper.addEditEntries(l);
+          }
+        }
+      } else {
+        skippedEdits ++;
+      }
+    }
+    ctx.getMetrics().incrLogEditsFiltered(skippedEdits);
+
+    return true;
+  }
+
+  @Override public UUID getPeerUUID() {
+    return this.ctx.getClusterId();
+  }
+
+  @Override public boolean canReplicateToSameCluster() {
+    return true;
+  }
+
+  @Override protected WALEntryFilter getScopeWALEntryFilter() {
+    // we do not care about scope. We replicate everything.
+    return null;
+  }
+
+  @Override public void start() {
+    startAsync();
+  }
+
+  @Override public void stop() {
+    // Stop shippers
+    for (Map.Entry<String, CatalogReplicaShipper[]> e : primaryRegionsUnderReplication.entrySet()) {
+      for (CatalogReplicaShipper s : e.getValue()) {
+        s.setRunning(false);
+      }
+    }
+    stopAsync();
+  }
+
+  @Override protected void doStart() {
+    notifyStarted();
+  }
+
+  @Override protected void doStop() {
+    notifyStopped();
+  }
+
+  public int getNumRetries() {
+    return numRetries;
+  }
+
+  public long getOperationTimeoutNs() {
+    return operationTimeoutNs;
+  }
+
+  private static class CatalogReplicaShipper extends Thread {
+    private final CatalogReplicaReplicationEndpoint endpoint;
+    private LinkedBlockingQueue<List<Entry>> queue;
+    private final AsyncClusterConnection conn;
+    //Indicates whether this particular worker is running
+    private volatile boolean isRunning = true;
+    private HRegionLocation location = null;
+    private RegionInfo regionInfo;
+    private long operationTimeoutNs;
+    private AsyncTableRegionLocator locator;
+
+    CatalogReplicaShipper(final CatalogReplicaReplicationEndpoint endpoint,
+      final AsyncClusterConnection connection, final HRegionLocation loc,
+      final RegionInfo regionInfo) {
+      this.endpoint = endpoint;
+      this.conn = connection;
+      this.location = loc;
+      this.regionInfo = regionInfo;
+      this.queue = new LinkedBlockingQueue<>();
+      this.operationTimeoutNs = endpoint.getOperationTimeoutNs();
+      locator = conn.getRegionLocator(regionInfo.getTable());
+    }
+
+    private void cleanQueue(final long startFlushSeqOpId) {
+      // remove entries with seq# less than startFlushSeqOpId
+
+      // TODO: It assumes one family in the queue.
+      // In the future, if more families are involved, it needs to have families into
+      // consideration.
+      queue.removeIf(n -> (n.get(0).getKey().getSequenceId() < startFlushSeqOpId));
+    }
+
+    public void addEditEntries(final List<Entry> entries) {
+      try {
+        if ((entries.size() == 1) && entries.get(0).getEdit().isMetaEdit()) {
```

Comment:
We'll never get a meta edit in a bunch of entries?

Reply:
Yeah, meta edit is always its own edit.